debian/bsondump.1
.\" Man page generated from reStructuredText.
.
.TH "BSONDUMP" "1" "January 30, 2015" "3.0" "mongodb-manual"
.SH NAME
bsondump \- MongoDB BSON Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
The \fBbsondump\fP converts \fIBSON\fP files into human\-readable
formats, including \fIJSON\fP\&. For example, \fBbsondump\fP is useful
for reading the output files generated by \fBmongodump\fP\&.
.sp
\fBIMPORTANT:\fP
.INDENT 0.0
.INDENT 3.5
\fBbsondump\fP is a diagnostic tool for inspecting
BSON files, not a tool for data ingestion or other application use.
.UNINDENT
.UNINDENT
.SH OPTIONS
.sp
Changed in version 3.0.0: \fBbsondump\fP removed the \fB\-\-filter\fP option.

.INDENT 0.0
.TP
.B bsondump
.UNINDENT
.INDENT 0.0
.TP
.B bsondump
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBbsondump\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs the \fBbsondump\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
output from \fIdatabase commands\fP
.IP \(bu 2
replication activity
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBbsondump\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-objcheck
Validates each \fIBSON\fP object before outputting it in \fIJSON\fP
format. By default, \fBbsondump\fP enables \fI\-\-objcheck\fP\&.
For objects with a high degree of sub\-document nesting,
\fI\-\-objcheck\fP can have a small impact on performance. You can set
\fI\-\-noobjcheck\fP to disable object checking.
.sp
Changed in version 2.4: MongoDB enables \fI\-\-objcheck\fP by default, to prevent any
client from inserting malformed or invalid BSON into a MongoDB
database.

.UNINDENT
.INDENT 0.0
.TP
.B \-\-noobjcheck
New in version 2.4.

.sp
Disables the default document validation that MongoDB performs on all
incoming BSON documents.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-type <=json|=debug>
Changes the operation of \fBbsondump\fP from outputting
"\fIJSON\fP" (the default) to a debugging format.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-pretty
New in version 3.0.0.

.sp
Outputs documents in a pretty\-printed format JSON.
.UNINDENT
.INDENT 0.0
.TP
.B <bsonFilename>
The final argument to \fBbsondump\fP is a document containing
\fIBSON\fP\&. This data is typically generated by
\fBbsondump\fP or by MongoDB in a \fIrollback\fP operation.
.UNINDENT
.SH USE
.sp
By default, \fBbsondump\fP outputs data to standard output. To
create corresponding \fIJSON\fP files, you will need to use the
shell redirect. See the following command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bsondump collection.bson > collection.json
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Use the following command (at the system shell) to produce debugging
output for a \fIBSON\fP file:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bsondump \-\-type=debug collection.bson
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
MongoDB Documentation Project
.SH COPYRIGHT
2011-2015
.\" Generated by docutils manpage writer.
.


debian/changelog
mongodb (2.5.4) unstable; urgency=low

  * see http://jira.mongodb.org/browse/SERVER/fixforversion/12898

 -- Ernie Hershey <ernie.hershey@mongodb.com>  Thu, 19 Dec 2013 00:41:46 -0500

mongodb (2.1.2) unstable; urgency=low

  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10894

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jun 2012 16:56:28 -0500

mongodb (1.9.2) unstable; urgency=low

  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Thu, 11 Aug 2011 16:56:28 -0500

mongodb (1.9.1) unstable; urgency=low

  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jul 2011 16:56:28 -0500

mongodb (1.9.0) unstable; urgency=low

  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10232

 -- Richard Kreuter <richard@10gen.com>  Sun, 01 May 2011 16:56:28 -0500

mongodb (1.8.0-rc0) unstable; urgency=low

  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10245

 -- Richard Kreuter <richard@10gen.com>  Tue, 22 Feb 2011 16:56:28 -0500

mongodb (1.7.6) unstable; urgency=low

  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10223

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongodb (1.7.5) unstable; urgency=low

  * durability journalingin place
  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10214

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongodb (1.7.4) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10210

 -- Richard Kreuter <richard@10gen.com>  Mon, 20 Dec 2010 16:56:28 -0500

mongodb (1.7.3) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10195

 -- Richard Kreuter <richard@10gen.com>  Tue, 16 Nov 2010 16:56:28 -0500

mongodb (1.7.2) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10194

 -- Richard Kreuter <richard@10gen.com>  Fri, 29 Oct 2010 16:56:28 -0500

mongodb (1.7.1) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongodb.org/browse/SERVER/fixforversion/10193

 -- Richard Kreuter <richard@10gen.com>  Tue, 28 Sep 2010 16:56:28 -0500

mongodb (1.7.0) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 03 Sep 2010 16:56:28 -0500

mongodb (1.5.8) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Tue, 03 Aug 2010 16:56:28 -0500

mongodb (1.5.7) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 30 Jul 2010 16:56:28 -0500


mongodb (1.5.6) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongodb.org/browse/SERVER/fixforversion/10179

 -- Richard Kreuter <richard@10gen.com>  Sat, 24 Jul 2010 16:56:28 -0500

mongodb (1.5.5) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongodb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 16 Jul 2010 16:56:28 -0500

mongodb (1.5.4) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongodb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 2 Jul 2010 16:56:28 -0500

mongodb (1.5.3) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongodb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Thu, 17 Jun 2010 16:56:28 -0500

mongodb (1.5.2) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongodb.org/browse/SERVER/fixforversion/10143

 -- Richard Kreuter <richard@10gen.com>  Wed, 27 May 2010 16:56:28 -0500

mongodb (1.5.1) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongodb.org/browse/SERVER/fixforversion/10142

 -- Richard Kreuter <richard@10gen.com>  Wed, 3 May 2010 16:56:28 -0500

mongodb (1.5.0) unstable; urgency=low

  * replication w & real-time, see http://jira.mongodb.org/browse/SERVER/fixforversion/10125

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500


mongodb (1.3.5) unstable; urgency=low

  * bug fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500

mongodb (1.3.4) unstable; urgency=low

  * bufg fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 17 Mar 2010 16:56:28 -0500

mongodb (1.3.3) unstable; urgency=low

  * geo

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongodb (1.3.2) unstable; urgency=low

  * munged debian files

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongodb (1.3.1) unstable; urgency=low

  * Initial release

 -- Kristina Chodorow <kristina@10gen.com>  Tue, 07 Apr 2009 10:18:58 -0400



src/mongo/gotools/bsondump/main/bsondump.go
// Main package for the bsondump tool.
package main

import (
	"github.com/mongodb/mongo-tools/bsondump"
	"github.com/mongodb/mongo-tools/common/db"
	"github.com/mongodb/mongo-tools/common/log"
	"github.com/mongodb/mongo-tools/common/options"
	"github.com/mongodb/mongo-tools/common/signals"
	"github.com/mongodb/mongo-tools/common/util"
	"os"
)

func main() {
	// initialize command-line opts
	opts := options.New("bsondump", bsondump.Usage, options.EnabledOptions{})
	bsonDumpOpts := &bsondump.BSONDumpOptions{}
	opts.AddOptions(bsonDumpOpts)

	args, err := opts.ParseArgs(os.Args[1:])
	if err != nil {
		log.Logvf(log.Always, "error parsing command line options: %v", err)
		log.Logvf(log.Always, "try 'bsondump --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	// print help, if specified
	if opts.PrintHelp(false) {
		return
	}

	// print version, if specified
	if opts.PrintVersion() {
		return
	}

	log.SetVerbosity(opts.Verbosity)
	signals.Handle()

	if len(args) > 1 {
		log.Logvf(log.Always, "too many positional arguments: %v", args)
		log.Logvf(log.Always, "try 'bsondump --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	// If the user specified a bson input file
	if len(args) == 1 {
		if bsonDumpOpts.BSONFileName != "" {
			log.Logvf(log.Always, "Cannot specify both a positional argument and --bsonFile")
			os.Exit(util.ExitBadOptions)
		}

		bsonDumpOpts.BSONFileName = args[0]
	}

	dumper := bsondump.BSONDump{
		ToolOptions:     opts,
		BSONDumpOptions: bsonDumpOpts,
	}

	reader, err := bsonDumpOpts.GetBSONReader()
	if err != nil {
		log.Logvf(log.Always, "Getting BSON Reader Failed: %v", err)
		os.Exit(util.ExitError)
	}
	dumper.BSONSource = db.NewBSONSource(reader)
	defer dumper.BSONSource.Close()

	writer, err := bsonDumpOpts.GetWriter()
	if err != nil {
		log.Logvf(log.Always, "Getting Writer Failed: %v", err)
		os.Exit(util.ExitError)
	}
	dumper.Out = writer
	defer dumper.Out.Close()

	log.Logvf(log.DebugLow, "running bsondump with --objcheck: %v", bsonDumpOpts.ObjCheck)

	if len(bsonDumpOpts.Type) != 0 && bsonDumpOpts.Type != "debug" && bsonDumpOpts.Type != "json" {
		log.Logvf(log.Always, "Unsupported output type '%v'. Must be either 'debug' or 'json'", bsonDumpOpts.Type)
		os.Exit(util.ExitBadOptions)
	}

	var numFound int
	if bsonDumpOpts.Type == "debug" {
		numFound, err = dumper.Debug()
	} else {
		numFound, err = dumper.JSON()
	}

	log.Logvf(log.Always, "%v objects found", numFound)
	if err != nil {
		log.Logv(log.Always, err.Error())
		os.Exit(util.ExitError)
	}
}
